keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,343 @@
+
+### User Stories
+
+#### Story 1: Scale up as fast as possible
+
+This mode is essential when you want to respond to a traffic increase quickly.
+
+Create an HPA with the following configuration:
+
+- `constraints`:
+  - `scaleUp`:
+    - `percent = 900`    (i.e., to increase the number of pods 10 times per minute is ok).
+
+All other parameters are not specified (default values are used)
+
+If the application is started with 1 pod, it will scale up with the following number of pods:
+
+    1 -> 10 -> 100 -> 1000
+
+So, it can reach `maxReplicas` very fast.
+
+Scale down will be done a usual way (check stabilization window in the [Stabilization Window][] section below and the [Algorithm details][] in the official HPA documentation)
+
+[Stabilization Window]: #stabilization-window
+[Algorithm details]: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### Story 2: Scale up as fast as possible, scale down very gradually
+
+This mode is essential when you don’t want to risk scaling down very quickly.
+
+Create an HPA with the constraints:
+
+- `constraints`:
+  - `scaleUp`:
+    - `percent = 900` (i.e. increase number of pods 10 times per minute is ok).
+  - `scaleDown`:
+    - `pods = 1`
+    - `periodSeconds = 600` (i.e., scale down one pod every 10 min)
+
+All other parameters are not specified (default values are used)
+
+If the application is started with 1 pod, it will scale up with the following number of pods:
+
+    1 -> 10 -> 100 -> 1000
+
+So, it can reach `maxReplicas` very fast.
+
+Scaling down will be by one pod each 10 min:
+
+    1000 -> 1000 -> 1000 -> … (7 more min) -> 999
+
+#### Story 3: Scale up very gradually, usual scale down process
+
+This mode is essential when you want to increase capacity, but you want it to be very pessimistic.
+
+Create an HPA with the constraints:
+
+- `constraints`:
+  - `scaleUp`:
+    - `pods = 1`    (i.e., increase only by one pod per minute)
+
+All other parameters are not specified (default values are used)
+
+If the application is started with 1 pod, it will scale up very gradually:
+
+    1 -> 2 -> 3 -> 4
+
+Scale down will be done a usual way (check stabilization window in [Algorithm details][])
+
+[Algorithm details]: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### Story 4: Scale up as usual, do not scale down
+
+This mode is essential when you don’t want to risk scaling down at all.
+
+Create an HPA with the following constraints:
+
+- `constraints`:
+  - `scaleDown`:
+    - `percent = 0`
+    - `pods = 0`
+
+i.e., set both constraints to 0, so that the HPA controller would never scale the cluster down
+
+All other parameters are not specified (default values are used)
+
+The cluster will scale up as usual (default values), but will never scale down.
+
+### Implementation Details/Notes/Constraints
+
+#### Algorithm pseudocode
+
+The algorithm to find the number of pods will look like this:
```

Comment:
We have found that customers care more about the rate of starts and stops (flapping) than the range over which those starts and stops oscillate.

For example, you might set `ScaleUpPods=10` per 10 minutes because there is a relatively high initialization cost when starting new pods.  The algorithm as described thus far is allowed to start up to 10 pods every minute if it also stops 10 pods every minute (oscillating within the 10 pod range).

I propose to instead implement scale up/down pods/percent as a "flap budget".  So in the above case HPA would only be able to start 10 pods before it exhausts its 10 minute budget.  This would rate limit the changes more precisely than just giving bounds.

Reply:
I have several points against this idea:
- My KEP is about adding configuration parameters to control scale velocity. Budget for any change seems to be quite a complex concept that should be thoroughly considered. I do understand why it might be needed. But I'm not sure I'm ready to consider all different use cases how scaling up/down with absolute/relative speed interact with different budgets.
- Flapping can be handled quite easily by setting strict constraints for scaling down. Like, `scaleUpPods=10 per 10min` and `scaleDownPercent=10 per 10min`. So you won't be able to remove more than 10% of pods for the same time interval.
- Flapping can also be fixed by increasing the stabilization window.

So, my overall recommendation would be not to over-complicate current KEP, it is already 8 pages long in my browser :)

Reply:
Your "histeresis" idea below also solve that problem.

Reply:
> you won't be able to remove more than 10% of pods for the same time interval

We might be talking about the same thing.  Let me ask a few clarifying questions:

Consider this scenario:
- currentReplicaCount=100
- scaleUpPods=10 per 10 min
- scaleDownPercent=10 per 10 min
- hpa sync interval=1 min

And every 1 minute the metric oscillates between high and low values such that the HPA is computing desiredReplicaCount: 110, 90, 110, 90, 110, 90,...

Would it be the case that desiredReplicaCount is updated every minute?  If yes, over the course of 10 minutes, you would accumulate 100 scale ups.  And 100% scale downs.  That would be a problem.

Reply:
Oh, I see your idea now.
Does it covered by the hysteresis described above?

Reply:
You're referring to this section, right? https://github.com/kubernetes/enhancements/blob/fe36d8a8699beb19bd7fab423bc12c213dd28cb2/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md#hpa-controller-state-changes

In particular, this sentence sounds like it captures the "flap budget" algorithm:

> if 30 seconds ago the number of replicas was increased by one, and we forbid to scale up for more than 1 pod per minute, then during the next 30 seconds, we won’t scale up again

So yes, I think this is good.